zmon-controller-app/src/main/java/org/zalando/zmon/service/impl/Grafana.java

```diff
@@ -0,0 +1,156 @@
+                    .queryParam("tag", params.containsKey("tag") ? URLEncoder.encode(params.get("tag"), "UTF-8") : "")
+                    .queryParam("limit", params.containsKey("limit") ? params.get("limit") : "25")
+                    .build();
+            Request request = Request.Get(url.toUri());
+            request.addHeader("Authorization", "Bearer " + token);
+            HttpResponse response = executor.execute(request).returnResponse();
+            return toResponseEntity(response);
+        } catch (Exception ex) {
+            log.error("Search grafana dashboard failed", ex);
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @Override
+    public ResponseEntity<JsonNode> upsertDashboard(String dashboard, String token) {
+        log.info("Creating/Updating grafana dashboard: user={}", authService.getUserName());
+        final Executor executor = Executor.newInstance(visualizationProperties.getHttpClient());
+        final String url = visualizationProperties.getUrl() + upsertDashboardEndpoint;
+
+        try {
+            Request request = Request.Post(url);
+            request.addHeader("Authorization", "Bearer " + token);
+            HttpResponse response = executor.execute(request.bodyString(
+                    dashboard, ContentType.APPLICATION_JSON))
+                    .returnResponse();
+            return toResponseEntity(response);
+        } catch (Exception ex) {
+            log.error("Create/Update grafana dashboard failed", ex);
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @Override
+    public ResponseEntity<JsonNode> deleteDashboard(String uid, String token) {
+        log.info("Deleting grafana dashboard: uid={} user={}", uid, authService.getUserName());
+
+        final Executor executor = Executor.newInstance(visualizationProperties.getHttpClient());
+        final String url = visualizationProperties.getUrl() + getAndDeleteDashboardEndPoint + uid;
+
+        try {
+            Request request = Request.Delete(url);
+            request.addHeader("Authorization", "Bearer " + token);
+            HttpResponse response = executor.execute(request).returnResponse();
+            return toResponseEntity(response);
+        } catch (Exception ex) {
+            log.error("Delete grafana dashboard: {} failed", uid, ex.getMessage());
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    private ResponseEntity<JsonNode> toResponseEntity(HttpResponse response) throws IOException {
+        int status = response.getStatusLine().getStatusCode();
+        HttpEntity entity = response.getEntity();
+        Integer allowedStatusCodes[] = {200, 412};
+
+        if (new HashSet<>(Arrays.asList(allowedStatusCodes)).contains(status) && entity != null) {
```

Comment:
you've created array, then passed it to list, and then passed it hash set. Not sure if it's optimal. Could you try to use Google Guava's `Sets.newHashSet` static method?